### PR TITLE
Elaborate getaddrinfo() test, add mock server to actually receive and verify DNS query sent by getaddrinfo() call

### LINKED ISSUE #15193: tests/net/socket/getaddrinfo tests too little
**Describe the bug**
Assert logic in https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/net/socket/getaddrinfo/src/main.c#L23 is backwards.  You can run this test without access to a local dnsmasq server and it passes.

**To Reproduce**
1. Make sure QEMU net-tools are started, **but no local dnsmasq server is available**.
2. cd tests/net/socket/getaddrinfo
3. mkdir build && cd build
4. cmake -DBOARD=qemu\_x86 ..
5. make run
6. test passes(?!?)

**Expected behavior**
Test should fail if no local dnsmasq server is running

**Impact**
Test is not working

**Screenshots or console output**
```
SeaBIOS (version rel-1.12.0-0-ga698c8995f-prebuilt.qemu.org)
Booting from ROM..***** Booting Zephyr OS v1.14.0-rc3-297-g6fcf986e7775 *****
Running test suite socket_getaddrinfo
===================================================================
starting test - test_getaddrinfo_ok
PASS - test_getaddrinfo_ok
===================================================================
starting test - test_getaddrinfo_no_host
PASS - test_getaddrinfo_no_host
===================================================================
Test suite socket_getaddrinfo succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Environment (please complete the following information):**
 - OS: Arch Linux
 - Toolchain: Zephyr SDK 0.10.0
 - SHA: https://github.com/zephyrproject-rtos/zephyr/commit/0734b4fcaaea62e5f887d932a79b31507d61429d

### PR DESCRIPTION
**Pull request by jukkar**
Fixes #15193

### CODE REVIEW

File: tests/net/socket/getaddrinfo/prj.conf [hunk 1]
```diff
@@ -15,19 +15,48 @@ CONFIG_TEST_RANDOM_GENERATOR=y
...
+#CONFIG_LOG_IMMEDIATE=y
+
+# We do not need neighbor discovery etc for this test
+CONFIG_NET_IPV6_DAD=n
+CONFIG_NET_IPV6_ND=n
+CONFIG_NET_IPV6_MLD=n
+
+# Disable internal ethernet drivers as the test is self contained
+# and does not need the on board driver to function.
+CONFIG_ETH_NATIVE_POSIX=n
```
Comment (andrewboie) [inline tests/net/socket/getaddrinfo/prj.conf]: this doesn't really scale as more drivers are added, is there another way?

File: tests/net/socket/getaddrinfo/prj.conf [hunk 2]
```diff
@@ -15,19 +15,48 @@ CONFIG_TEST_RANDOM_GENERATOR=y
...
+CONFIG_ETH_MCUX=n
+CONFIG_ETH_SAM_GMAC=n
+CONFIG_ETH_DW=n
+CONFIG_ETH_ENC28J60=n
+CONFIG_ETH_STM32_HAL=n
+
+# The test does not work with userspace
+CONFIG_TEST_USERSPACE=n
```
Comment (andrewboie) [inline tests/net/socket/getaddrinfo/prj.conf]: NACK!!

This test used to, hence the calls to "ztest_user_unit_test" in the test case.
If getaddrinfo is broken in user mode, at minimum a medium-priority bug needs to be open on it..

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: tests/net/socket/getaddrinfo/src/main.c [hunk 3]
```diff
@@ -47,6 +250,7 @@ void test_main(void)
 	k_thread_system_pool_assign(k_current_get());
 
 	ztest_test_suite(socket_getaddrinfo,
+			 ztest_user_unit_test(test_getaddrinfo_setup),
```
Comment (andrewboie) [inline tests/net/socket/getaddrinfo/src/main.c]: if this is the only scenario that doesn't work in user mode, instantiate it with "ztest_unit_test" instead..

File: tests/net/socket/getaddrinfo/prj.conf (hunk 2 above)
Comment (jukkar) [reply tests/net/socket/getaddrinfo/prj.conf]: getaddrinfo() is working just fine in userspace, but for the test application I used k_sem which caused issues, so by disabling userspace I was able to run the test app.

File: tests/net/socket/getaddrinfo/prj.conf (hunk 1 above)
Comment (jukkar) [reply tests/net/socket/getaddrinfo/prj.conf]: I do not think this is a big issue as we currently run the test app in qemu_x86 and native_posix. I think we can remove these settings if needed, just need to make sure there is no ill effects. I just added these in order to make sure that only loopback interface will ever be created in the application.

File: tests/net/socket/getaddrinfo/src/main.c (hunk 3 above)
Comment (jukkar) [reply tests/net/socket/getaddrinfo/src/main.c]: Ah, I did not know that. Thanks for the info, will try that then.

Comment (andrewboie) [reply tests/net/socket/getaddrinfo/src/main.c]: @jukkar it seems you have done this for other tests, disabling userspace completely for tests/net/getnameinfo even though it works in user mode.

Please do not do this again without consulting me, disabling user mode tests for APIs which are supposed to be tested under user mode will just lead to bit-rot and unwelcome surprises later.

Comment (andrewboie) [reply tests/net/socket/getaddrinfo/src/main.c]: Also, I'm seeing there are new socket APIs without system calls. I'm fixing what I can, but please reject in code review any new BSD Socket API implementations that don't work in user mode.

Comment (jukkar) [reply tests/net/socket/getaddrinfo/src/main.c]: Yep, good point, will do.

File: tests/net/socket/getaddrinfo/prj.conf (hunk 1 above)
Comment (andrewboie) [reply tests/net/socket/getaddrinfo/prj.conf]: ok

Comment (jukkar) [reply tests/net/socket/getaddrinfo/prj.conf]: Removed the disabling of board specific ethernet drivers, the tests work just fine without them.

File: tests/net/socket/getaddrinfo/prj.conf (hunk 2 above)
Comment (jukkar) [reply tests/net/socket/getaddrinfo/prj.conf]: Removed the disabling of userspace, only two tests are now run in non-userspace mode.

File: tests/net/socket/getaddrinfo/src/main.c [hunk 4]
```diff
@@ -47,7 +259,9 @@ void test_main(void)
 	k_thread_system_pool_assign(k_current_get());
 
 	ztest_test_suite(socket_getaddrinfo,
-			 ztest_user_unit_test(test_getaddrinfo_ok),
+			 ztest_unit_test(test_getaddrinfo_setup),
+			 ztest_unit_test(test_getaddrinfo_ok),
```
Comment (andrewboie) [inline tests/net/socket/getaddrinfo/src/main.c]: why is test_getaddrinfo_ok no longer running in user mode?
all it does is call getaddrinfo() and take a semaphore.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: tests/net/socket/getaddrinfo/src/main.c (hunk 4 above)
Comment (jukkar) [reply tests/net/socket/getaddrinfo/src/main.c]: I will convert it to use sys_mutex so that the test case works in user mode.

**Comment by jukkar** [comment]
recheck

**Comment by pfalcon** [comment]
> As the getaddrinfo() test is suppose to be self contained, then
we will need to install handler for the sent DNS query. After we
receive the query in process_dns() thread, we verify that the
query was sent properly. We do not return DNS reply back to
the caller in this case

I appreciate all the effort went into implementing this, and the commit message which describes these details, but I'd say that the source code also should contain comments (e.g. function comments) explaining that it implements mock, receive-only DNS server to catch and verify a query sent by the getaddrinfo() call. (Because otherwise my first reaction was "what all this stuff does in a test called  tests/net/socket/getaddrinfo" ?)

**Comment by pfalcon** [comment]
I think "net: core: In RX check if the pkt is for loopback interface" commit should be submitted separately and merged as a bugfix for 1.14 regardless if the getaddrinfo test updates will get thru into the release (I hope they will).

**Comment by jukkar** [comment]
> commit should be submitted separatel

Not really needed as the fix is only needed for dummy L2 which is only used in tests.

**Comment by jukkar** [comment]
No need to have this on 1.14, so marking it for 2.0

**Comment by jukkar** [comment]
As I said, no need to backport this to 1.14 as the original test (although inadequate) works just fine -> removing backport label.

**Comment by mike-scott** [comment]
> but I'd say that the source code also should contain comments (e.g. function comments) explaining that it implements mock, receive-only DNS server to catch and verify a query sent by the getaddrinfo() call. 

IMHO, this patch looks good to me as is:
- From a patch perspective, there is a clear and defined purpose with that expressed in the description.
- From source code point of view: this is a test.  Not to be copied/pasted and used for samples.  The function naming + comments sprinkled through-out clearly explain why the code exists and what function it's serving.  I don't think we need to add more explanation, otherwise we're comments become white-noise.

Here's a clear example of a comment in the code saying exactly what you are asking for:
```
/* In this test we are just checking if the query came to us in correct
 * form, we are not creating a DNS server implementation here.
 */
```

**Review by mike-scott — APPROVED** [review]

**Review by pfalcon — APPROVED** [review]
> I'd say that the source code also should contain comments (e.g. function comments) explaining that it implements mock, receive-only DNS server to catch and verify a query sent by the getaddrinfo() call.

Let me repeat this suggestion, but otherwise giving +1.

**Comment by jukkar** [comment]
ping @andrewboie, can you review?

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**